### PR TITLE
Allow user to customize Attribute comparison by supplying a *cmpspec*

### PR DESCRIPTION
**Pull request by botant**
I've implemented the initial part of issue #435. I wanted to stop at this point and get some feedback in case you guys think I'm going in the wrong direction.

The main idea of how this feature would be used can be found in ``test_cmp.py``.

I did the basics of the check list below, and I tried my best to update the ``.pyi`` but I'm not sure whether what I did is correct. If this goes through, I'll update the documentation once we have all features down.

- [x] Added **tests** for changed code.
- [**?**] New features have been added to our [Hypothesis testing strategy](https://github.com/python-attrs/attrs/blob/master/tests/strategies.py).
- [x] Changes or additions to public APIs are reflected in our type stubs (files ending in ``.pyi``).
    - [**not yet**] ...and used in the stub test file `tests/typing_example.py`.
- [**not yet**] Updated **documentation** for changed code.
    - [**not yet**] New functions/classes have to be added to `docs/api.rst` by hand.
    - [x] Changes to the signature of `@attr.s()` have to be added by hand too.
    - [x] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded).  Find the appropriate next version in our [``__init__.py``](https://github.com/python-attrs/attrs/blob/master/src/attr/__init__.py) file.
- [**not yet**] Documentation in `.rst` files is written using [semantic newlines](https://rhodesmill.org/brandon/2012/one-sentence-per-line/).
- [x] Changes (and possible deprecations) have news fragments in [`changelog.d`](https://github.com/python-attrs/attrs/blob/master/changelog.d).

### CODE REVIEW

**Comment by jriddy** [comment]
Does doing this at the class level make sense? It's my intuition that the problems are individual attributes, which can be of types that have ~ill-conceived~ non-standard behavior of the comparison methods.

This is an honest question I have; I don't know the right answer. It may just be too unergonomic to specify comparison types on each attribute, and then too difficult to efficiently generate a comparison method from that.

**Comment by botant** [comment]
Hi Josh,

Apologies, maybe I got the naming convention mixed up and the text of the
PR is not what I meant.

It is at the level of individual attributes, not the class decorated with
attr.s( ).

Something like this:

@attr.s()
class MyClass:
   my_att = attr.ib(cmpspec=MySpec)


Cheers!
Antonio

**Comment by jriddy** [comment]
Okay awesome! This will form a good base for a more ergonomic API then!

I think the target API for this should be something very similar to how validators and default providers work: they give you the choice of a `keyword=provider` method or a decorator based `@attribute.keyword` method. And I don't think this would be hard with what you've done, something like...

```python
@attr.s
class ColorspacedImage:
    space: str = attr.ib()
    data: np.ndarray = attr.ib()

    @data.comparator
    class _NdCmp:
        def eq(self, inst, other):
            return (inst == other).all()
        def ne(self, inst, other):
            return not self.eq(inst, other)
        ...
```

I don't think we need to support all that now, but I could see that as a goal and direction to take this in.

I'll start working on a PR to this to show what I mean.

**Comment by botant** [comment]
Yes, that is what I had in mind as well, something consistent with
attr.validators and attr.converters.

The ``@data.comparator`` decorator is a great idea, especially together
with ``@functools.total_ordering``.

File: src/attr/_make.py [hunk 1]
```diff
@@ -1190,52 +1210,43 @@ def __ne__(self, other):
...
+        """
+        return _attrs_to_cmp_tuple(obj, attrs)
+
     # We can't just do a big self.x = other.x and... clause due to
     # irregularities like nan == nan is false but (nan,) == (nan,) is true.
-    if attrs:
-        lines.append("    return  (")
-        others = ["    ) == ("]
-        for a in attrs:
-            lines.append("        self.%s," % (a.name,))
```
Comment (jriddy) [inline src/attr/_make.py]: I think we should strive to keep this generated code behavior here if possible. It looks like a bit of work and tuning was put into getting a bunch of details right (for speed, debugging, serialization), and I don't want to throw that out to solve and edgier edge case. I have an idea, i'll try it in my coming PR

**Comment by jriddy** [comment]
I will also say that the more I look at this, the less I like "cmpspec" as a word for this concept. It's just clumsy and it doesn't make a pleasant sound in my head. How do you feel about "comparator"?

**Comment by botant** [comment]
Yes, cmpspec doesn't exactly roll of the tongue, does it? I don't like comparator either, mainly because it isn't a word (outside electronics) but at least it is in line with validator and converter...

**Comment by jriddy** [comment]
Okay I'll toss it around in my head, see if I can't come up with something
we can all like better. I would have suggested just using `cmp` and
triggering the advanced behavior on whether that attribute is a class, but
they've deprecated that attribute and I think that might be confusing now.

Which reminds me... I'm thinking we should also define the way that this
feature interacts with `eq` and `order`, with my gut telling me to just
raise an error if you try to use this rich comparison class and those are
not `None`.

**Comment by botant** [comment]
I had a different take on the interaction between `comparator` and `eq` + `order`. We have already been through one round of deprecation from `cmp` to `eq` + `order`, so maybe we should avoid a second one.

How about keeping `eq` and `order` exactly as they are now, i.e. their values determine whether a given attribute is included in the dunder comparison methods, whilst `comparator` provides the *how*?

**Comment by jriddy** [comment]
Hmm that actually makes sense, especially if you think of providing default
comparators for some common use cases (numpy, pandas, etc)

**Pull request closed without merging**